### ISSUE TITLE
Fatal Error using Visual R Markdown on Rstudio 1.4.727

### ISSUE THREAD

**Issue by njtierney** [rstudio/rstudio#7631]
### System details

    RStudio Edition : Desktop
    RStudio Version : 1.4.727
    OS Version      : MacOS Mojave 10.14.6 (18G6020)
    R Version       : 4.0.2

### Steps to reproduce the problem

1. Download + install Rstudio 1.4.727 (Also occurs in Rstudio version 1.4.722 - does not occur in 1.3.1073 as the visual rmarkdown is not available)
1. Launch Rstudio
1. Navigate to Rmarkdown document
1. Click the  ![image](https://user-images.githubusercontent.com/6488485/90855352-c49cea80-e3c2-11ea-8ab8-39ebfd9d30fd.png) button 
1. rstudio aborts

### Describe the problem in detail

When I click the ![image](https://user-images.githubusercontent.com/6488485/90855352-c49cea80-e3c2-11ea-8ab8-39ebfd9d30fd.png) button I get a fatal error, and RStudio aborts:

![image](https://user-images.githubusercontent.com/6488485/90855434-f0b86b80-e3c2-11ea-80c1-6c2d117d4a77.png)

I cannot open the rstudio project again because it then tries to launch the visual mode again and crashes once more

### Describe the behavior you expected

To view the visual mode without error.

- [X] I have read the guide to submitting good bug reports at https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports . 
- [X] I have installed the latest version of RStudio and confirmed that the issue still persists.
- [X] If I am reporting a RStudio crash, I have included a diagnostics report. https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report

[diagnostics-report.txt](https://github.com/rstudio/rstudio/files/5106968/diagnostics-report.txt)

- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by jjallaire** [comment]
Daily build v1.4.734 has a _speculative_ fix for this issue. LMK if it prevents the problem on your end.

**Comment by njtierney** [comment]
Hi @jjallaire 

> When you open Console, is there anything that appears to be RStudio related (either the rsession or RStudio process) in the User Reports pane?

I don't get anything in the **User Reports** pane, just some old issues with `phantomjs` and **Excel** that I had a little while ago.

I confirm that I get the same report in **system.log** as before:

```
Aug 24 09:39:39 MU00104512X rsession-ntie0001[88599]: ERROR Unexpected exception: Null value not allowed for this type; LOGGED FROM: int rstudio::r::session::RReadConsole(const char *, CONSOLE_BUFFER_CHAR *, int, int) src/cpp/r/session/RStdCallbacks.cpp:402
Aug 24 09:39:39 MU00104512X rsession-ntie0001[88599]: ERROR R SUICIDE: Unexpected exception: Null value not allowed for this type; LOGGED FROM: void (anonymous namespace)::rSuicide(const std::string &) src/cpp/session/SessionMain.cpp:1163
```

> Another thing to try, do you get the crash if you run with RStudio v1.4.712:

I do not get a crash with `v1.4.713:` actually! This works fine. Interesting.

> If so, do you also get the crash with v1.4.713:

For completeness, no I don't a crash with `v1.4.713`

> Another question: do you have a version of pandoc installed at the system level (you could test this by going to a terminal and typing `which pandoc`).

Yes, I do have `pandoc`:

```
~/github/njtierney$ which pandoc
/usr/local/bin/pandoc
~/github/njtierney$ pandoc -v
pandoc 2.10.1
Compiled with pandoc-types 1.21, texmath 0.12.0.2, skylighting 0.8.5
Default user data directory: /Users/ntie0001/.local/share/pandoc or /Users/ntie0001/.pandoc
Copyright (C) 2006-2020 John MacFarlane
Web:  https://pandoc.org
This is free software; see the source for copying conditions.
There is no warranty, not even for merchantability or fitness
for a particular purpose.
```

> Try running the following sequence of commands at the terminal (from any directory):

```
touch filter.lua
touch test.md
/Applications/RStudio.app/Contents/MacOS/pandoc/pandoc -L filter.lua test.md -o test.html
``` 

(For this I went back to the problematic version of RStudio, `v1.4.727`, let me know if you want me to try with another version)

I get no output from this:

```
~/github/njtierney$ touch filter.lua
~/github/njtierney$ touch test.md
~/github/njtierney$ /Applications/RStudio.app/Contents/MacOS/pandoc/pandoc -L filter.lua test.
md -o test.html
~/github/njtierney$ 
```

Thanks once again for your patience, hopefully this yielded some useful info. Let me know how I can help from here.

**Comment by njtierney** [comment]
Just saw your comment on using the rstudio version 1.4.734

I saw on the daily builds it has 1.4.735, installing that, I got a fatal error, with the following from **Console**:

```
Aug 24 15:44:03 MU00104512X rsession-ntie0001[24485]: ERROR Unexpected exception: Null value not allowed for this type; LOGGED FROM: int rstudio::r::session::RReadConsole(const char *, CONSOLE_BUFFER_CHAR *, int, int) src/cpp/r/session/RStdCallbacks.cpp:402
Aug 24 15:44:03 MU00104512X rsession-ntie0001[24485]: ERROR R SUICIDE: Unexpected exception: Null value not allowed for this type; LOGGED FROM: void (anonymous namespace)::rSuicide(const std::string &) src/cpp/session/SessionMain.cpp:1163
```

**Comment by jjallaire** [comment]
It's _fantastic_ news that you don't get the crash with v1.4.713 as that means:

1) It has nothing to do with pandoc v2.10.1 (that was the first version that included the pandoc upgrade)

2) I believe we now have inevitability on this bug :-)

If you can bear with me a bit we should be able to solve this pretty easily. The next thing to do is to binary search through the versions from v1.4.713 to v1.4.727 (so e.g. try out v1.4.720, if that works fine, then try out v1.4.723, etc.) until we narrow in on the exact version where the crash was introduced. That should give us a range of commits from which the crash originated, hopefully we can then combine this w/ the info that (a) The error is an "Unexpected Null value" exception; and (b) The error appears (from your first screenshot) to be occurring while we probe to see if there is a project-level bibliography for the document.

Incidentally, does this crash happen for all Rmd files entering visual mode or just the README.Rmd from your screenshot?

**Comment by njtierney** [comment]
OK fantastic, looking forward to applying a binary search! We'll get this thing soon. :)

Just to confirm because I feel a bit vulnerable editing URLs to get a download link, do I just edit the file name here to match the version:

`s3.amazonaws.com/rstudio-ide-build/desktop/macos/RStudio-1.4.{xxx}.dmg`

Or is there another place they are listed?

The crash happened initially on another project, not just the README.Rmd I was using from the screenshot.

**Comment by jjallaire** [comment]
Yes, that's correct re: editing the URL. Here is where they are listed: https://dailies.rstudio.com/rstudio/oss/mac/

Let's see what the binary search produces then as necessary determine if the crash is triggered by specific file contents or configuration.

**Comment by ronblum** [comment]
I don't think there's much I can add, but @jjallaire would you still like to see what happens on Mojave?

**Comment by kevinushey** [comment]
It looks like this error message ("Null value not allowed for this type.") might be coming from SOCI:

https://github.com/SOCI/soci/blob/68d4a8aea84277e948b6cf956e4533b0d8f85115/include/soci/unsigned-types.h#L27-L30

If that's indeed the case, we might want to ask if this rings any bells for @kfeinauer. I'm not sure what we'd be doing when switching to Visual Mode that would trigger something here, though.

**Comment by kfeinauer** [comment]
I believe that means that we got some data back from SOCI for a particular column which was null. See http://soci.sourceforge.net/doc/release/4.0/indicators/.

In this case, we attempted to cast that to unsigned char as part of some query, but an exception is thrown because we are not specifically handling nulls here. 

As far as fixing: either disallow nulls on the column in question, or pass a `soci::indicator` into the `withOutput` call in question. (Note: have not tested to see how this actually works - may need to experiment a little with it).

**Comment by jjallaire** [comment]
Bingo! @njtierney My expectation then is that v1.4.721 will work fine and v1.4.722 will have the crash (this is when we merged a bunch of stuff related to Zotero, and our Zotero integration uses SOCI to read the Zotero local SQLLite database).

@njtierney I'm assuming you have Zotero installed? If so it would be ideal if we could get a zipped copy of your Zotero Data Directory (typically `~/Zotero`). We understand though if you aren't comfortable sharing this publicly. In that case you could email to `jj@rstudio.com`. If you don't want to do that then we can just try speculative fixes until we get it.

**Comment by jjallaire** [comment]
@ronblum This isn't going to be OS related so no need to try on Mojave.

**Comment by jjallaire** [comment]
@dragonstyle has a speculative fix here: https://github.com/rstudio/rstudio/commit/9c870e0cfe5a15d5ebc25b57377fa8f6a4665adb

@njtierney I'll ping this thread back when there is a binary with the fix available. In the meantime it would still be helpful for us to get your `~/Zotero` directory if you are comfortable sharing it (_completely_ understand though if you don't want to).

**Comment by njtierney** [comment]
@jjallaire  I'll send you an email with the zotero directory! I'll try out `v1.4.721` and `v1.4.722` later today and send an update :)

**Comment by jjallaire** [comment]
Here is the speculative fix (v1.4.737) for the Zotero null field issue: https://s3.amazonaws.com/rstudio-ide-build/desktop/macos/RStudio-1.4.737.dmg

Let us know if that solves the problem 🤞 

Thanks also for the Zotero dir, we'll try to repro with that as well.

**Comment by njtierney** [comment]
OK, alas, this did not work:

```
Aug 25 12:02:44 MU00104512X rsession-ntie0001[55585]: ERROR Unexpected exception: Null value not allowed for this type; LOGGED FROM: int rstudio::r::session::RReadConsole(const char *, CONSOLE_BUFFER_CHAR *, int, int) src/cpp/r/session/RStdCallbacks.cpp:402
Aug 25 12:02:44 MU00104512X rsession-ntie0001[55585]: ERROR R SUICIDE: Unexpected exception: Null value not allowed for this type; LOGGED FROM: void (anonymous namespace)::rSuicide(const std::string &) src/cpp/session/SessionMain.cpp:1163
```

Would you like me to try the binary search?

**Comment by jjallaire** [comment]
Okay, your Zotero database provided what we needed (thanks for that!). It turns out that you had an empty table that we were doing a `SELECT(MAX)` on, and for sqlite that fittingly returns `NULL` rather than 0. This fix does indeed resolve the problem seen w/ your database: https://github.com/rstudio/rstudio/commit/5fac189dc73bfbc0d38dbfbd8310ddc481007afe

A daily build with this fix should be available within a couple of hours (will ping back when it is).

**Comment by jjallaire** [comment]
@njtierney I believe that this build should fix the issue: https://s3.amazonaws.com/rstudio-ide-build/desktop/macos/RStudio-1.4.739.dmg

LMK how this goes on your end!

**Comment by njtierney** [comment]
It works! 🎉 

Some notes: I had some strange error message (something to do with `rstudio_function::boost` - sorry I can't get more descriptive than that. Unfortunately, I closed RStudio before I captured it), but this I think was a holdover from the previous repo (`brolgar`), which had previously had a fatal error. I noticed this because I also opened up another rstudio project that I hadn't opened before and it worked no problems, and also once I ran `rm -rf .Rproj.user` on the `brolgar` project, Visual R Markdown worked no problems.

So, I think that's case closed?

Thanks again for your help and patience with this, visual markdown mode has been a daily driver for me and it's been a blast.
